### PR TITLE
fix: If internal links are not specified, the external one should be used

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this PR do?
If internal links are not specified, the external one should be used

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/18643

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

Signed-off-by: Oleksandr Andriienko <oandriie@redhat.com>

### CODE REVIEW

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java [hunk 1]
```diff
@@ -29,12 +30,15 @@
...
     this.cheServerEndpoint = cheServerEndpoint;
   }
 
   @Override
   public Pair<String, String> get(RuntimeIdentity runtimeIdentity) throws InfrastructureException {
-    return Pair.of(CHE_API_INTERNAL_VARIABLE, cheServerEndpoint);
+    if (this.cheServerEndpoint != null) {
+      return Pair.of(CHE_API_INTERNAL_VARIABLE, cheServerEndpoint);
+    }
+    return null;
```
Comment (metlos) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java:42]: This breaks https://github.com/eclipse-che/che-server/blob/07263f1e30089689d71b057f747a44a29283e3c4/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/provision/env/CheApiEnvVarProvider.java#L48

while that env var is deprecated, we are still providing it to the workspace pods as far as I am aware.

**Review by metlos — COMMENTED** [review]
Looks good at a glance, apart from the one unhandled NPE mentioned below.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java (hunk 1 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java:42]: CHE_API_VARIABLE !=che.api.internal @metlos can you explain how it's related?

Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java:42]: The current behavior is that che server endpoint is never null and the `get` method of the `KubernetesCheApiInternalEnvVarProvider` never returns a `null` value. This is the assumption in `CheApiEnvVarProvider` which uses the return value from this method without a null check.

By making the `KubernetesCheApiInternalEnvVarProvider` potentially return `null`, we're breaking that assumption and open the possibility of an NPE. I have no idea if that is possible practically, just see the possibility in the code.

Comment (skabashnyuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java:42]: I think the whole point of this pr is that internal can be null. If some code needs something that never can be null it should use the results of CheApiEnvVarProvider.java right?

Comment (AndrienkoAleksandr) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java:42]: +1 for @skabashnyuk conclusions.

Comment (AndrienkoAleksandr) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java:42]: And I think we have null check https://github.com/eclipse-che/che-server/pull/45/files#diff-a433d052bb9fbe17e92d0cac528935ba2dfae6c982ce08dcd95c0b23b26403a1R52. I see CheApiEnvVarProvider is deprecated. So ideally not null value api endpoint should provide CheApiExternalEnvVarProvider like successor. This provider should provide public url. So logic is simple for application which used api env variables: if internal url env variable is absent or empty, than should be used value from external public url env variable. See pr: https://github.com/eclipse-che/che-theia/pull/1161. Also for che-theia I removed usage deprecated CheApiEnvVarProvider in this pr.

Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/KubernetesCheApiInternalEnvVarProvider.java:42]: That null-check was what I was after. Thanks for adding it :)

**Review by tolusha — APPROVED** [review]

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**